src/lib.rs

```diff
@@ -150,110 +213,260 @@ impl<'object, Endian> DebugInfo<'object, Endian>
+                programs: vec![],
+                phantom: PhantomData,
+            }
         };
-        let comp_dir = entry.attr(gimli::DW_AT_comp_dir)
-            .and_then(|attr| attr.string_value(debug_str));
-        let comp_name = entry.attr(gimli::DW_AT_name)
-            .and_then(|attr| attr.string_value(debug_str));
-
-        Some(Unit {
-            address_size: header.address_size(),
-            ranges: ranges,
-            line_offset: line_offset,
-            comp_dir: comp_dir,
-            comp_name: comp_name,
-        })
+
+        // Do we also need to extract function information?
+        if func_names.is_none() {
+            return Some(unit);
+        }
+
+        while let Ok(Some((_, entry))) = entries.next_dfs() {
+            // We only care about functions
+            match entry.tag() {
+                gimli::DW_TAG_inlined_subroutine |
+                gimli::DW_TAG_subprogram => (),
+                _ => continue,
+            }
+
+            // Where does this function live?
+            let ranges = Self::get_ranges(entry, debug_ranges, header.address_size());
+            if ranges.is_empty() {
+                continue;
+            }
+
+            if cfg!(debug_assertions) {
+                // All programs should be fully contained within their compilation unit
+                for r in &ranges {
+                    assert!(unit.ranges
+                        .iter()
+                        .any(|range| range.begin <= r.begin && range.end >= r.end));
+                }
+            }
+
+            // For naming, we prefer the linked name, if available
+            if let Some(name) = entry.attr(gimli::DW_AT_linkage_name)
+                .and_then(|attr| attr.string_value(debug_str)) {
+                unit.programs.push(Program {
+                    ranges: ranges,
+                    inlined: entry.tag() == gimli::DW_TAG_inlined_subroutine,
+                    name: name,
+                });
+                continue;
+            }
+
+            // If we don't have the link name, check if this function refers to another
+            if let Some(gimli::AttributeValue::UnitRef(origin)) =
+                entry.attr_value(gimli::DW_AT_abstract_origin) {
+                if let Some(name) = func_names.as_ref().unwrap().get(&origin.0) {
+                    unit.programs.push(Program {
+                        ranges: ranges,
+                        inlined: entry.tag() == gimli::DW_TAG_inlined_subroutine,
+                        name: name,
+                    });
+                    continue;
+                } else {
+                    panic!("failed to resolve subprogram origin {:x}", entry.offset().0);
+                }
+            }
+
+            // Linked name is not available, and we don't have a reference to another function.
+            // Fall back to just plain old name, if that's available.
+            if let Some(name) = entry.attr(gimli::DW_AT_name)
```

Comment:
All this naming stuff is getting pretty involved -- would be nice to factor it out into smaller functions.